### PR TITLE
validation: Improve input script check error reporting

### PR DESCRIPTION
**Pull request by dergoegge**
An input script might be invalid for multiple reasons. For example, it might fail both a standardness check and a consensus check, which can lead to a `mandatory-script-verify-flag-failed` error being reported that includes the script error string from the standardness failure (e.g. `mandatory-script-verify-flag-failed (Using OP_CODESEPARATOR in non-witness script)`), which is confusing.

### CODE REVIEW

**Comment by darosior** [comment]
utACK f859ff8a4e9c3aa23bf5be6eceb7099ca72b2290

**Comment by darosior** [comment]
Here is a functional test which creates a transaction using an `OP_CODESEPARATOR` in its `scriptSig` (non-standard) but also uses an `OP_RETURN` right after, which makes it invalid by consensus. It returns the invalid combination of error strings fixed in this PR ("mandatory-script-verify-flag-failed" because it's a consensus failure, but "Using OP_CODESEPARATOR in non-witness script" because it returns the error string for the first check).

You should be able to trivially adapt it to the fix in this PR by updating the error message.

```diff
diff --git a/test/functional/data/invalid_txs.py b/test/functional/data/invalid_txs.py
index 33054fd517..6c259d59bb 100644
--- a/test/functional/data/invalid_txs.py
+++ b/test/functional/data/invalid_txs.py
@@ -245,6 +245,19 @@ class TooManySigops(BadTxTemplate):
             script_pub_key=lotsa_checksigs,
             amount=1)
 
+
+class NonStandardAndInvalid(BadTxTemplate):
+    """A non-standard transaction which is also consensus-invalid should return the consensus error."""
+    reject_reason = "mandatory-script-verify-flag-failed (Using OP_CODESEPARATOR in non-witness script)"
+    expect_disconnect = True
+    valid_in_block = False
+
+    def get_tx(self):
+        return create_tx_with_script(
+            self.spend_tx, 0, script_sig=b'\x00' * 3 + b'\xab\x6a',
+            amount=(self.spend_avail // 2))
+
+
 def getDisabledOpcodeTemplate(opcode):
     """ Creates disabled opcode tx template class"""
     def get_tx(self):
diff --git a/test/functional/p2p_invalid_tx.py b/test/functional/p2p_invalid_tx.py
index 0ae05d4b0b..7782217152 100755
--- a/test/functional/p2p_invalid_tx.py
+++ b/test/functional/p2p_invalid_tx.py
@@ -165,7 +165,7 @@ class InvalidTxRequestTest(BitcoinTestFramework):
             node.p2ps[0].send_txs_and_test([rejected_parent], node, success=False)
 
         self.log.info('Test that a peer disconnection causes erase its transactions from the orphan pool')
-        with node.assert_debug_log(['Erased 100 orphan transaction(s) from peer=25']):
+        with node.assert_debug_log(['Erased 100 orphan transaction(s) from peer=26']):
             self.reconnect_p2p(num_connections=1)
 
         self.log.info('Test that a transaction in the orphan pool is included in a new tip block causes erase this transaction from the orphan pool')
```

**Comment by dergoegge** [comment]
@darosior added the test to this PR, thanks!

File: src/validation.cpp [hunk 1]
```diff
@@ -2192,6 +2194,14 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
                         flags & ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS, cacheSigStore, &txdata);
                 if (check2())
                     return state.Invalid(TxValidationResult::TX_NOT_STANDARD, strprintf("non-mandatory-script-verify-flag (%s)", ScriptErrorString(check.GetScriptError())));
+
+                // If the second check failed, it failed due to a mandatory script verification
+                // flag, but the first check might have failed on a non-mandatory script
+                // verification flag.
```
Comment (ariard) [inline src/validation.cpp:2200]: I was reading the comment just above this one i.e “Check whether the failure was caused by a non-mandatory script verification check, such as non-standard DER encodings or non-null dummy arguments”.

I think the 2 examples given in this comment are actually “mandatory-script-verify-flag”, i.e CONSENSUS and not NOT_STANDARD:
- DER encodings (BIP66 and `SCRIPT_VERIFY_DERSIG`), interpreter flag is set in `GetBlockScriptFlags` https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2378
- NULLDUMMY for CHECKMULTISIG (BIP147 and `SCRIPT_VERIFY_NULLDUMMY`), interpreter flag is set in `GetBlockScriptFlags` https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2393

Those 2 consensus changes have been buried, so I think the comment could reflet there are TX_NOT_STANDARD before the buried height (correspondingly `363725` and `481824`) and verification check error for DER encoding and `NULLDUMMY` should be TX_CONSENSUS after.

File: test/functional/data/invalid_txs.py [hunk 2]
```diff
@@ -263,6 +263,17 @@ def get_tx(self):
         'valid_in_block' : True
         })
 
+class NonStandardAndInvalid(BadTxTemplate):
+    """A non-standard transaction which is also consensus-invalid should return the consensus error."""
+    reject_reason = "mandatory-script-verify-flag-failed (OP_RETURN was encountered)"
+    expect_disconnect = True
+    valid_in_block = False
+
+    def get_tx(self):
```
Comment (ariard) [inline test/functional/data/invalid_txs.py:272]: I think we should have further tests that a non-standard transaction which is consensus valid before an activation height, is considered as consensus invalid after the activation height.

E.g on regtest, the deployment height are always active unless overridden. I don’t think we do that in invalid_txs.py` currently: https://github.com/bitcoin/bitcoin/blob/master/test/functional/p2p_invalid_tx.py#L28

**Review by ariard — COMMENTED** [review]
utACK ad51775

File: src/validation.cpp [hunk 3]
```diff
@@ -2192,6 +2194,14 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
                         flags & ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS, cacheSigStore, &txdata);
                 if (check2())
                     return state.Invalid(TxValidationResult::TX_NOT_STANDARD, strprintf("non-mandatory-script-verify-flag (%s)", ScriptErrorString(check.GetScriptError())));
```
Comment (ariard) [inline src/validation.cpp:2196]:
```
diff --git a/src/validation.cpp b/src/validation.cpp
index fc61b14dc1..ecc1ab22dd 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1970,7 +1970,7 @@ bool CheckInputScripts(const CTransaction& tx, TxValidationState& state,
                 CScriptCheck check2(txdata.m_spent_outputs[i], tx, i,
                         flags & ~STANDARD_NOT_MANDATORY_VERIFY_FLAGS, cacheSigStore, &txdata);
                 if (check2())
-                    return state.Invalid(TxValidationResult::TX_NOT_STANDARD, strprintf("non-mandatory-script-verify-flag (%s)", ScriptErrorString(check.GetScriptError())));
+                    return state.Invalid(TxValidationResult::TX_NOT_STANDARD, strprintf("non-mandatory-script-verify-flag (%s)", ScriptErrorString(error)));
 
                 // If the second check failed, it failed due to a mandatory script verification
                 // flag, but the first check might have failed on a non-mandatory script
```

**Review by ariard — COMMENTED** [review]
Re-tested, getting the correct RPC string.

**Comment by dergoegge** [comment]
I don't know what to do about the CI failure. I guess we need a fix for #30960? In the mean time I can remove the test commit.

**Comment by dergoegge** [comment]
> -par=1 (or whatever to make the thread inline) should work, no?

That worked, thanks!

File: src/validation.cpp (hunk 3 above)
Comment (dergoegge) [reply src/validation.cpp:2196]: Will do if i push again

**Review by darosior — COMMENTED** [review]
re-ACK 86e2a6b749c7fecbd086b361806ac9f6e9426d79

**Review by ariard — COMMENTED** [review]
Re-Code Review ACK 86e2a6b7

The code nit improvement, ready to be git cherry-picked:https://github.com/ariard/bitcoin/tree/2024-10-check2-impr

**Review by instagibbs — APPROVED** [review]
ACK 86e2a6b749c7fecbd086b361806ac9f6e9426d79

played with tests to ensure the test was failing how I'd expect it with modifications

File: src/validation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply src/validation.cpp:2196]: @sipa might be able to pull this into #31112, given it's following up.

**Review by danielabrozzoni — COMMENTED** [review]
post merge tACK 86e2a6b749c7fecbd086b361806ac9f6e9426d79

**Pull request merged**